### PR TITLE
8276227: ciReplay: SIGSEGV if classfile for replay compilation is not present after JDK-8275868

### PR DESCRIPTION
**Pull request by chhagedorn**
The fix for [JDK-8275868](https://bugs.openjdk.java.net/browse/JDK-8275868) does not handle the case when the classfile for the method to be replay compiled is not present. This will fail to load the klass. Afterwards, we are trying to access the protection domain of the failed to load klass (i.e. a null pointer) which results in a segmentation fault. The fix is straight forward to only set the new protection domain if the klass was loaded successfully. I additionally changed the code such that we are only trying to set the protection domain when reading the first `instanceKlass` entry. This avoids some potential problems with older replay files where we do not have this additional first entry set by JDK-8275868.

Thanks,
Christian

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276227](https://bugs.openjdk.java.net/browse/JDK-8276227): ciReplay: SIGSEGV if classfile for replay compilation is not present after JDK-8275868

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 4fe95743205abbfbe7a03306ebeb911499f1621a
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to 4fe95743205abbfbe7a03306ebeb911499f1621a

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6189/head:pull/6189` \
`$ git checkout pull/6189`

Update a local copy of the PR: \
`$ git checkout pull/6189` \
`$ git pull https://git.openjdk.java.net/jdk pull/6189/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6189`

View PR using the GUI difftool: \
`$ git pr show -t 6189`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6189.diff">https://git.openjdk.java.net/jdk/pull/6189.diff</a>

</details>

### CODE REVIEW

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review! Unfortunately, none of our tests caught this. I observed this when I did some additional testing and had not specified the correct class paths. I've added a test which covers this.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by dean-long — APPROVED** [review]

File: src/hotspot/share/ci/ciReplay.cpp [hunk 1]
```diff
@@ -874,12 +876,18 @@ class CompileReplay : public StackObj {
...
       // The first entry is the holder class of the method for which a replay compilation is requested.
       // Use the same protection domain to load all subsequent classes in order to resolve all classes
       // in signatures of inlinees. This ensures that inlining can be done as stated in the replay file.
       _protection_domain = Handle(_thread, k->protection_domain());
     }
+
+    // Only initialize the protection domain handle with the protection domain of the very first entry.
+    // This also ensures that older replay files work.
+    _protection_domain_initialized = true;
+
```
[the remark this answers is not in the document]
Comment (dean-long) [reply src/hotspot/share/ci/ciReplay.cpp:890]: I'm having second thoughts on supporting old replay files.  It's easy enough to add a version number, which allows us to introduce incompatible changes without breaking old replay files.  I'll probably introduce a version number with my fix for 8276095.

[the remark this answers is not in the document]
Comment (chhagedorn) [reply src/hotspot/share/ci/ciReplay.cpp:890]: You're right. A version number would solve this completely. This implementation is more robust than the previous one but not complete. Here we only try to set the protection domain once where in the previous implementation, we would have picked the first non-null protection domain found (which could happen after looking at many classes). Do you want to revisit this code with the introduction of version numbers in 8276095 and we proceed with this temporary fix?

[the remark this answers is not in the document]
Comment (dean-long) [reply src/hotspot/share/ci/ciReplay.cpp:890]: Yes, go ahead with your fix.

[the remark this answers is not in the document]
Comment (chhagedorn) [reply src/hotspot/share/ci/ciReplay.cpp:890]: Great, thanks Dean for your review!

**Pull request closed without merging**